src/daemon/metric.c

```diff
@@ -0,0 +1,675 @@
+/**
+ * collectd - src/daemon/metric.c
+ * Copyright (C) 2019-2020  Google LLC
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a
+ * copy of this software and associated documentation files (the "Software"),
+ * to deal in the Software without restriction, including without limitation
+ * the rights to use, copy, modify, merge, publish, distribute, sublicense,
+ * and/or sell copies of the Software, and to permit persons to whom the
+ * Software is furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ *
+ * Authors:
+ *   Florian octo Forster <octo at collectd.org>
+ *   Manoj Srivastava <srivasta at google.com>
+ **/
+
+#include "collectd.h"
+
+#include "metric.h"
+#include "plugin.h"
+
+/* Label names must match the regex `[a-zA-Z_][a-zA-Z0-9_]*`. Label names
+ * beginning with __ are reserved for internal use.
+ *
+ * Source:
+ * https://prometheus.io/docs/concepts/data_model/#metric-names-and-labels */
+#define VALID_LABEL_CHARS                                                      \
+  "abcdefghijklmnopqrstuvwxyz"                                                 \
+  "ABCDEFGHIJKLMNOPQRSTUVWXYZ"                                                 \
+  "0123456789_"
+/* Metric names must match the regex `[a-zA-Z_:][a-zA-Z0-9_:]*` */
+#define VALID_NAME_CHARS VALID_LABEL_CHARS ":"
+
+int value_marshal_text(strbuf_t *buf, value_t v, metric_type_t type) {
+  switch (type) {
+  case METRIC_TYPE_GAUGE:
+  case METRIC_TYPE_UNTYPED:
+    return strbuf_printf(buf, GAUGE_FORMAT, v.gauge);
+  case METRIC_TYPE_COUNTER:
+    return strbuf_printf(buf, "%" PRIu64, v.counter);
+  default:
+    ERROR("Unknown metric value type: %d", type);
```

Comment:
At this point we don't know what the value type is, and we can't be sure that it is an int. 
The printf function uses the format specifiers to figure out what to pop off the stack. So when it sees %d, it pops off 4 bytes and interprets them as an int, which is wrong (the binary representation of (float)3.0 is not the same as (int)3).

You'll need to either use the %f format specifiers or cast the arguments to int.

Reply:
Well no, but actually yes.

This line is not printing the value, but the value's type. `type` is a `metric_type_t`, which is an `enum`. [Section 6.7.2.2](http://c0x.coding-guidelines.com/6.7.2.2.html) of the C11 standard states:

> The  identifiers  in  an  enumerator  list  are  declared  as  constants  that  have  type `int` and may   appear   wherever   such are permitted.

*and*

> Each enumerated type shall be compatible with char, a signed integer type, or an unsigned integer type.

If I read this correctly, then `size(METRIC_TYPE_GAUGE) == 4` because the *identifier*, i.e. `METRIC_TYPE_GAUGE`, is treated like an `int`. However, `sizeof((metric_type_t){METRIC_TYPE_GAUGE})`, which allocates a `metric_type_t` on the stack and takes its size, may yield other sizes, e.g. 1 or 2 bytes, because the *type*, i.e. `metric_type_t`, may have different storage sizes.

I've added a cast to `int` to be on the save side, though I think that compilers will likely warn about a problem caused by printf format and argument mismatch.